prototype2/multiblade/mbcommon/MBConfig.cpp

```diff
@@ -0,0 +1,91 @@
+/* Copyright (C) 2018 European Spallation Source, ERIC. See LICENSE file */
+//===----------------------------------------------------------------------===//
+///
+/// \file
+///
+/// \brief using nlohmann json parser to read configurations from file
+//===----------------------------------------------------------------------===//
+
+#include <common/Log.h>
+#include <fstream>
+#include <multiblade/mbcommon/MBConfig.h>
+#include <nlohmann/json.hpp>
+
+///
+MBConfig::MBConfig(std::string jsonfile) : ConfigFile(jsonfile) {
+
+  loadConfigFile();
+
+  if (instrument == InstrumentGeometry::Estia) {
+    detector = new MB16Detector(Digitisers);
+  }
+
+  assert(detector != nullptr);
+}
+
+///
+void MBConfig::loadConfigFile() {
+  nlohmann::json root;
+
+  if (ConfigFile.empty()) {
+    LOG(INIT, Sev::Info, "JSON config - no config file specified, using default configuration");
+    return;
+  }
+
+  LOG(INIT, Sev::Info, "JSON config - loading configuration from file {}", ConfigFile);
```

Comment:
If the goal is to always return from the function without throwing, then you should likely wrap the following ifstream-related statements in a try-catch block as well?

Reply:
yes. I actually hate the whole section of repeated try{} catch {} sequences, but it seems like the json parser throws on just about every error condition. This makes it difficult to provide meaningful error messages to the user regarding which field is in error. Any suggestions for doing this differently?

Reply:
Some ideas:
* don't catch. Let it throw and be caught higher up, thus failing. Run with debugger to trace the exact line the exception came from.
* if you want to log and move on, you could maybe just include the entire json string in the exception. You could also then have a standalone utility for validating a config that someone can subsequently run (again in debug mode) to find the exact problem.
Just off the top of my head. Not saying these are best practices or anything.